### ISSUE TITLE
Would you accept a provider based on native Win32 APIs?

### ISSUE THREAD

**Issue by thomcc** [sqlcipher/sqlcipher#450]
On Windows, for various reasons it can be painful to build and bundle third-party cryptography libraries like OpenSSL, NSS, or LibTomCrypt.

As of Windows 7[^winver], it's completely possible to support SQLcipher's needs by using the functionality in the [Windows CNG APIs](https://learn.microsoft.com/en-us/windows/win32/seccng/cng-portal) (e.g. `bcrypt.{lib,dll}`), which is shipped and updated with the operating system.

That is, if a provider were written based on this, it would allow SQLcipher users on Windows to avoid having to build/link/distribute a third party cryptography library, since they'd just be using the "native" crypto API on the platform (similar to how macOS/iOS users can use CommonCrypto). I don't mind contributing this (and have started on it [here](https://github.com/thomcc/sqlcipher/tree/bcrypt), although it is still WIP), and maintaining it for you -- it would simplify my own open source code[^oss].

Unfortunately, I noticed that you've stopped accepting new providers. This makes sense, but I'm hoping you might make an exception for one based on native win32 APIs provided by the OS. So, is there any chance you'd accept it?

[^oss]: Concretely: I want this because it will simplify maintenance of [rusqlite](https://github.com/rusqlite/rusqlite), which has a fair amount of complexity from building and statically linking a copy of OpenSSL on Windows, mostly to support a use case where users want SQLcipher and don't want to manage any C dependencies themselves. If SQLcipher had a provider that used native windows APIs, I could migrate these users to that.

[^winver]: These APIs were added in Windows Vista, but SQLcipher would need at least Windows 7 (e.g. for `BCryptDeriveKeyPBKDF2`).

**Comment by thomcc** [comment]
> I can leave this ticket open for now, but would you mind if we circled back with you in the new year with a more definitive answer?

Yep, that sounds great to me. I agree that it's a lot of work, definitely I wouldn't suggest this if it weren't significantly more convenient for linking on windows.

**Comment by nmoore-genetec** [comment]
Is there any update on this ? I also would like the flexibility of deploying on Windows without the need for openssl

**Comment by alexeydott** [comment]
You might find it helpful to take a look at this fork; in it, the `sqlcipher_provider` interface is [implemented using CNG primitives](https://github.com/alexeydott/sqlcipher/blob/master/doc/sqlcipher-cng-provider.md). 
It’s not exactly “standard” SQLite (I added a few extensions I needed and ensured C89 compatibility in some places so that sqlcipher could be compiled for use with Delphi/CBuilder), but as far as bcrypt is concerned, everything works exactly as intended—at least for my needs.
